src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1344,676 +1431,434 @@ private bool IsMiniShell()
+            Dbg.Assert(command != null, "Caller should validate the parameter");
+            _command = command;
         }
 
         #endregion constructor
 
-        /// <summary>
-        /// Reader for output stream of process
-        /// </summary>
-        private ProcessStreamReader _outputReader;
-        /// <summary>
-        /// Reader for error stream of process
-        /// </summary>
-        private ProcessStreamReader _errorReader;
-        /// <summary>
-        /// Synchronized object queue in which object read form output and 
-        /// error streams are deposited
-        /// </summary>
-        private ObjectStream _processOutput;
+        private SteppablePipeline _pipeline;
+        private Serializer _xmlSerializer;
 
         /// <summary>
-        /// Start reading the output/error. Note all the work is done asynchronously.
+        /// Add an object to write to process
         /// </summary>
-        internal void Start()
+        /// <param name="input"></param>
+        internal void Add(object input)
         {
-            _processOutput = new ObjectStream(128);
+            if (_stopping || _streamWriter == null)
+            {
+                // if _streamWriter is already null, then we already called Done()
+                // so we should just discard the input.
+                return;
+            }
 
-            // Start async reading of error and output
-            // readercount variable is used by multiple threads to close "processOutput" ObjectStream.
-            // Without properly initializing the readercount, the ObjectStream might get
-            // closed early. readerCount is protected here by using the lock.
-            lock (_readerLock)
+            if (_inputFormat == NativeCommandIOFormat.Text)
             {
-                if (_redirectOutput)
-                {
-                    _readerCount++;
-                    _outputReader = new ProcessStreamReader(_process.StandardOutput, _processPath, true, _processOutput.ObjectWriter, this);
-                    _outputReader.Start();
-                }
-                if (_redirectError)
-                {
-                    _readerCount++;
-                    _errorReader = new ProcessStreamReader(_process.StandardError, _processPath, false, _processOutput.ObjectWriter, this);
-                    _errorReader.Start();
-                }
+                AddTextInput(input);
+            }
+            else // Xml
+            {
+                AddXmlInput(input);
             }
         }
 
-        /// <summary>
-        /// Stops reading from streams. This is called from NativeCommandProcessor's StopProcessing
-        /// method. Note return of this method doesn't mean reading has stopped and all threads are
-        /// done. 
-        /// Use Done to ensure that all reading threads are finished.
-        /// </summary>
-        internal void Stop()
+        private void AddTextInput(object input)
```

Comment:
This method is called from only one place and probably will be inlined by the compiler anyway. How about removing this method, and use `AddTextInputFromFormattedArray(_pipeline.Process(input));` directly in `Add(object input)`?